File: backend/app/schemas/inspection.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator

# ── Item-level result ──


class InspectionItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    record_id: int
    item_type: str
    success: bool
    value: str | None = None
    unit: str | None = None
    status: str  # normal / warning / critical / error
    details: dict | None = None
    raw_output: str | None = None
    error_message: str | None = None
    command_used: str | None = None
    executed_at: datetime | None = None
    duration_ms: int | None = None

    @field_serializer("executed_at")
    @classmethod
    def _serialize_dt(cls, v: datetime | None) -> str | None:
        if v is None:
            return None
        s = v.isoformat()
        if v.tzinfo is None:
            return s + "Z"
        return s


# ── Record-level (without items) ──


class InspectionRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    # PVE 虚拟机巡检记录 device_id 为 NULL(迁移 0036),必须允许 None,
    # 否则 model_validate 直接 ValidationError → 详情接口 500
    device_id: int | None = None
    device_name: str
    device_ip: str | None = None
    target_type: str
    # 巡检时留档的精确 OS 名(迁移 0037;历史记录为 None)
    os_name: str | None = None
    mode: str
    status: str
    total_items: int = 0
    normal_count: int = 0
    warning_count: int = 0
    critical_count: int = 0
    error_count: int = 0
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_ms: int | None = None
    triggered_by: int | None = None
    batch_id: str | None = None

    @field_serializer("started_at", "finished_at")
    @classmethod
    def _serialize_dt(cls, v: datetime | None) -> str | None:
        if v is None:
            return None
        s = v.isoformat()
        if v.tzinfo is None:
            return s + "Z"
        return s


# ── Record with items (detail view) ──


class InspectionRecordDetail(InspectionRecordResponse):
    items: list[InspectionItemResponse] = []


# ── Run request ──


class InspectionRunRequest(BaseModel):
    device_ids: list[int] = Field(min_length=1, max_length=500)
    mode: str = "standard"  # standard / custom / quick / core
    items: list[str] | None = None  # only for mode=custom
    timeout: int = Field(default=30, ge=1, le=300)
    username: str | None = Field(default=None, max_length=255)
    password: str | None = Field(default=None, max_length=4096)
    ssh_key: str | None = Field(default=None, max_length=16384)
    credential_id: int | None = None

    @field_validator("mode")
    @classmethod
    def validate_mode(cls, v):
        if v not in ("standard", "custom", "quick", "core"):
            raise ValueError("mode must be 'standard', 'custom', 'quick', or 'core'")
        return v

    @field_validator("device_ids")
    @classmethod
    def validate_device_ids(cls, v):
        if not v:
            raise ValueError("请至少选择一台设备")
        return v


# ── Paginated history response ──


class InspectionRecordListResponse(BaseModel):
    items: list[InspectionRecordResponse]
    total: int
    page: int
    page_size: int


# ── Items catalog (for custom mode UI) ──


class InspectionItemInfo(BaseModel):
    type: str
    label: str
    quick: bool = False  # whether included in quick mode


class InspectionItemsCatalogResponse(BaseModel):
    target_type: str
    items: list[InspectionItemInfo]


# ── 分指标告警阈值（前端按 item_type 查表配色）──
#
# items 的 key 是巡检项 item_type（cpu / memory / disk / failed_services），
# 不是指标页那套 cpu_pct / mem_pct / disk_max_pct 字段名——两套命名空间之间的映射
# 由前端负责。数值原样来自 app/services/inspection_parser.py 的 THRESHOLDS，
# 后端不在别处再存一份，免得同一个百分比又出现两处真相。


class MetricThreshold(BaseModel):
    warning: float
    critical: float


class MetricThresholdsResponse(BaseModel):
    items: dict[str, MetricThreshold]


# ── Report statistics ──


class InspectionReportResponse(BaseModel):
    total_inspections: int = 0
    total_items_checked: int = 0
    overall_health: dict = {}  # {normal_pct, warning_pct, critical_pct, error_pct}
    by_target_type: dict = {}  # {linux: {count, warning_rate, critical_rate}, ...}
    top_warnings: list[dict] = []
    trend: list[dict] = []

File: frontend/src/types/inspection.ts
// ── Inspection Types ──

export interface InspectionItemResult {
  id: number
  record_id: number
  item_type: string
  success: boolean
  value: string | null
  unit: string | null
  status: 'normal' | 'warning' | 'critical' | 'error'
  details: Record<string, any> | null
  raw_output: string | null
  error_message: string | null
  command_used: string | null
  executed_at: string | null
  duration_ms: number | null
}

export interface InspectionRecord {
  id: number
  // PVE 虚拟机巡检记录 device_id 为 NULL(迁移 0036)
  device_id: number | null
  device_name: string
  device_ip: string | null
  target_type: 'linux' | 'windows'
  // core：自动化运维的健康巡检固定使用的模式，检查项由后端按操作系统自动决定
  mode: 'standard' | 'custom' | 'quick' | 'core'
  status: 'running' | 'completed' | 'failed' | 'partial'
  total_items: number
  normal_count: number
  warning_count: number
  critical_count: number
  error_count: number
  started_at: string | null
  finished_at: string | null
  duration_ms: number | null
  triggered_by: number | null
  batch_id: string | null
}

export interface InspectionRecordDetail extends InspectionRecord {
  items: InspectionItemResult[]
}

export interface InspectionDeviceItem {
  id: number
  name: string
  ip_address: string | null
  type: string
  status: string
  os_system: string | null
  credential_id: number | null
  target_type: 'linux' | 'windows'
}

export interface InspectionItemInfo {
  type: string
  label: string
  quick: boolean
}

export interface InspectionRunRequest {
  device_ids: number[]
  mode: 'standard' | 'custom' | 'quick'
  items?: string[]
  timeout?: number
  username?: string
  password?: string
  credential_id?: number
}

export interface InspectionRunResult {
  batch_id: string | null
  results: InspectionDeviceResult[]
  total: number
  succeeded: number
  failed: number
}

export interface InspectionDeviceResult {
  // PVE 虚拟机巡检记录 device_id 为 NULL(迁移 0036)
  device_id: number | null
  device_name: string
  ip_address: string | null
  target_type: string
  record_id?: number
  status: string
  total_items: number
  normal_count: number
  warning_count: number
  critical_count: number
  error_count: number
  duration_ms: number
  items: InspectionItemResult[]
  error?: string
}

export interface InspectionReport {
  total_inspections: number
  total_items_checked: number
  overall_health: {
    normal_pct: number
    warning_pct: number
    critical_pct: number
    error_pct: number
  }
  by_target_type: Record<string, { count: number; warning_rate: number; critical_rate: number }>
  top_warnings: Array<{
    device_name: string
    target_type: string
    warning_count: number
    critical_count: number
    timestamp: string | null
  }>
  trend: Array<{
    date: string
    normal: number
    warning: number
    critical: number
    error: number
  }>
}

// ── 阈值表 ──
// 由 GET /api/inspection/thresholds 下发(权限 device:view),key 是巡检项 item_type
// (cpu / memory / disk / failed_services …),不是指标名 cpu_pct / mem_pct,映射由前端负责。

/** 单个巡检项的告警 / 严重阈值,边界一律按 >= 判定 */
export interface MetricThreshold {
  warning: number
  critical: number
}

export type MetricThresholdTable = Record<string, MetricThreshold>

export interface MetricThresholdsResponse {
  items: MetricThresholdTable
}
